File: bbgm/core/contract_negotiation.py
import random

from flask import g

from bbgm import app
from bbgm.core import play_menu
from bbgm.util import get_payroll, lock
import bbgm.util.const as c

def new(pid, resigning=False):
    """Start a new contract negotiation with player.

    Args: 
        pid: An integer that must correspond with a free agent.
        resigning: A boolean. True if this is a negotiation for a contract
            extension with a current player who just became a free agent. False
            otherwise.

    Returns:
        False if the new negotiation is started successfully. Otherwise, it
        returns a string containing an error message to be sent to the user.
    """
    app.logger.debug('Trying to start new contract negotiation with player %d' % (pid,))

    if resigning:
        resigning = True
    else:
        resigning = False

    if (g.phase >= c.PHASE_AFTER_TRADE_DEADLINE and g.phase <= c.PHASE_AFTER_DRAFT) and not resigning:
        return "You're not allowed to sign free agents now."
    r = g.dbex('SELECT COUNT(*) FROM player_attributes WHERE tid = :tid', tid=g.user_tid)
    num_players_on_roster, = r.fetchone()
    if num_players_on_roster >= 15 and not resigning:
        return "Your roster is full. Before you can sign a free agent, you'll have to buy out or release one of your current players.";
    if not lock.can_start_negotiation():
        return "You cannot initiate a new negotiaion while game simulation is in progress or a previous contract negotiation is in process."
    r = g.dbex('SELECT tid FROM player_attributes WHERE pid = :pid', pid = pid)
    if r.rowcount:
        tid, = r.fetchone()
        if tid != c.PLAYER_FREE_AGENT:
            return "Player %d is not a free agent." % (pid,)
    else:
        return "Player %d does not exist." % (pid,)

    # Initial player proposal
    r = g.dbex('SELECT contract_amount*(1+free_agent_times_asked/10), contract_expiration FROM player_attributes WHERE pid = :pid', pid = pid)
    player_amount, expiration = r.fetchone()
    player_years = expiration - g.season
    # Adjust to account for in-season signings
    if g.phase <= c.PHASE_AFTER_TRADE_DEADLINE:
        player_years += 1

    max_offers = random.randint(1, 5)

    g.dbex('INSERT INTO negotiation (pid, team_amount, team_years, player_amount, player_years, num_offers_made, max_offers, resigning) VALUES (:pid, :player_amount, :player_years, :player_amount, :player_years, 0, :max_offers, :resigning)', pid=pid, player_amount=player_amount, player_years=player_years, max_offers=max_offers, resigning=resigning)
    lock.set_negotiation_in_progress(True)
    play_menu.set_status('Contract negotiation in progress...')
    play_menu.refresh_options()

    # Keep track of how many times negotiations happen with a player
    if not resigning:
        g.dbex('UPDATE player_attributes SET free_agent_times_asked = free_agent_times_asked + 1 WHERE pid = :pid', pid = pid)

    return False

def offer(pid, team_amount, team_years):
    """Make an offer to a player.

    pid must correspond with an ongoing negotiation.
    """
    app.logger.debug('User made contract offer for %d over %d years to %d' % (team_amount, team_years, pid))

    if team_amount > 20000:
        team_amount = 20000
    if team_years > 5:
        team_years = 5
    if team_amount < 500:
        team_amount = 500
    if team_years < 1:
        team_years = 1

    r = g.dbex('SELECT player_amount, player_years, num_offers_made, max_offers FROM negotiation WHERE pid = :pid', pid = pid)
    player_amount, player_years, num_offers_made, max_offers = r.fetchone()

    num_offers_made += 1
    if num_offers_made <= max_offers:
        if team_years < player_years:
            player_years -= 1
            player_amount *= 1.2
        elif team_years > player_years:
            player_years += 1
            player_amount *= 1.2
        if team_amount < player_amount and team_amount > 0.7 * player_amount:
            player_amount = .75 * player_amount + .25 * team_amount
        elif team_amount < player_amount:
            player_amount *= 1.1
        if team_amount > player_amount:
            player_amount = team_amount
    else:
        player_amount = 1.05 * player_amount

    if player_amount > 20000:
        player_amount = 20000
    if player_years > 5:
        player_years = 5

    g.dbex('UPDATE negotiation SET team_amount = :team_amount, team_years = :team_years, player_amount = :player_amount, player_years = :player_years, num_offers_made = :num_offers_made WHERE pid = :pid', team_amount=team_amount, team_years=team_years, player_amount=player_amount, player_years=player_years, num_offers_made=num_offers_made, pid=pid)

def accept(pid):
    """Accept the player's offer.

    pid must correspond with an ongoing negotiation.

    Returns False if everything works. Otherwise, a string containing an error
    message (such as "over the salary cap") is returned.
    """
    app.logger.debug('User accepted contract proposal from %d' % (pid,))

    r = g.dbex('SELECT player_amount, player_years, resigning FROM negotiation WHERE pid = :pid', pid = pid)
    player_amount, player_years, resigning = r.fetchone()

    # If this contract brings team over the salary cap, it's not a minimum
    # contract, and it's not resigning a current player, ERROR!
    payroll = get_payroll(g.user_tid)
    if not resigning and (payroll + player_amount > g.salary_cap and player_amount != 500):
        return 'This contract would put you over the salary cap. You cannot go over the salary cap to sign free agents to contracts higher than the minimum salary. Either negotiate for a lower contract, buy out a player currently on your roster, or cancel the negotiation.'

    # Adjust to account for in-season signings
    if g.phase <= c.PHASE_AFTER_TRADE_DEADLINE:
        player_years -= 1

    r = g.dbex('SELECT MAX(roster_position) + 1 FROM player_attributes WHERE tid = :tid', tid = g.user_tid)
    roster_position, = r.fetchone()

    g.dbex('UPDATE player_attributes SET tid = :tid, contract_amount = :contract_amount, contract_expiration = :contract_expiration, roster_position = :roster_position WHERE pid = :pid', tid=g.user_tid, contract_amount=player_amount, contract_expiration=g.season + player_years, roster_position=roster_position, pid=pid)

    g.dbex('DELETE FROM negotiation WHERE pid = :pid', pid = pid)
    lock.set_negotiation_in_progress(False)
    play_menu.set_status('Idle')
    play_menu.refresh_options()

    return False

def cancel(pid):
    """Cancel contract negotiations with a player.

    pid must correspond with an ongoing negotiation.
    """
    app.logger.debug('User canceled contract negotiations with %d' % (pid,))

    # Delete negotiation
    g.dbex('DELETE FROM negotiation WHERE pid = :pid', pid = pid)

    # If no negotiations are in progress, update status
    r = g.dbex('SELECT 1 FROM negotiation')
    if r.rowcount == 0:
        lock.set_negotiation_in_progress(False)
        play_menu.set_status('Idle')
        play_menu.refresh_options()

def cancel_all():
    """Cancel all ongoing contract negotiations.
    As of the time that I'm writing this, the only time there should be multiple
    ongoing negotiations in the first place is when a user is resigning players
    at the end of the season, although that should probably change eventually.
    """
    app.logger.debug('Canceling all ongoing contract negotiations...')

    # If no negotiations are in progress, update status
    r = g.dbex('SELECT pid FROM negotiation')
    for pid, in r.fetchall():
        cancel(pid)

File: js/core/contractNegotiation.js
'use strict';

var g = require('../globals');
var ui = require('../ui');
var freeAgents = require('./freeAgents');
var player = require('./player');
var team = require('./team');
var Promise = require('bluebird');
var eventLog = require('../util/eventLog');
var helpers = require('../util/helpers');
var lock = require('../util/lock');

/**
 * Start a new contract negotiation with a player.
 *
 * @memberOf core.contractNegotiation
 * @param {IDBTransaction} tx An IndexedDB transaction on gameAttributes, messages, negotiations, and players, readwrite.
 * @param {number} pid An integer that must correspond with the player ID of a free agent.
 * @param {boolean} resigning Set to true if this is a negotiation for a contract extension, which will allow multiple simultaneous negotiations. Set to false otherwise.
 * @param {number=} tid Team ID the contract negotiation is with. This only matters for Multi Team Mode. If undefined, defaults to g.userTid.
 * @return {Promise.<string=>)} If an error occurs, resolve to a string error message.
 */
function create(tx, pid, resigning, tid) {
    tid = tid !== undefined ? tid : g.userTid;

    if ((g.phase >= g.PHASE.AFTER_TRADE_DEADLINE && g.phase <= g.PHASE.RESIGN_PLAYERS) && !resigning) {
        return Promise.resolve("You're not allowed to sign free agents now.");
    }

    // Can't flatten because of error callbacks
    return lock.canStartNegotiation(tx).then(function (canStartNegotiation) {
        if (!canStartNegotiation) {
            return "You cannot initiate a new negotiaion while game simulation is in progress or a previous contract negotiation is in process.";
        }

        return tx.players.index('tid').count(g.userTid).then(function (numPlayersOnRoster) {
            if (numPlayersOnRoster >= 15 && !resigning) {
                return "Your roster is full. Before you can sign a free agent, you'll have to release or trade away one of your current players.";
            }

            return tx.players.get(pid).then(function (p) {
                var negotiation, playerAmount, playerYears;

                if (p.tid !== g.PLAYER.FREE_AGENT) {
                    return p.name + " is not a free agent.";
                }

                // Initial player proposal;
                playerAmount = freeAgents.amountWithMood(p.contract.amount, p.freeAgentMood[g.userTid]);
                playerYears = p.contract.exp - g.season;
                // Adjust to account for in-season signings;
                if (g.phase <= g.PHASE.AFTER_TRADE_DEADLINE) {
                    playerYears += 1;
                }

                if (freeAgents.refuseToNegotiate(playerAmount, p.freeAgentMood[g.userTid])) {
                    return '<a href="' + helpers.leagueUrl(["player", p.pid]) + '">' + p.name + '</a> refuses to sign with you, no matter what you offer.';
                }

                negotiation = {
                    pid: pid,
                    tid: tid,
                    team: {amount: playerAmount, years: playerYears},
                    player: {amount: playerAmount, years: playerYears},
                    orig: {amount: playerAmount, years: playerYears},
                    resigning: resigning
                };

                return tx.negotiations.add(negotiation).then(function () {
                    require('../core/league').updateLastDbChange();
                    ui.updateStatus("Contract negotiation");
                    return ui.updatePlayMenu(tx);
                });
            });
        });
    });
}

/**
 * Cancel contract negotiations with a player.
 *
 * @memberOf core.contractNegotiation
 * @param {number} pid An integer that must correspond with the player ID of a player in an ongoing negotiation.
 * @return {Promise}
 */
function cancel(pid) {
    return g.dbl.tx(["gameAttributes", "messages", "negotiations"], "readwrite", function (tx) {
        return tx.negotiations.delete(pid).then(function () {
            // If no negotiations are in progress, update status
            return lock.negotiationInProgress(tx);
        }).then(function (negotiationInProgress) {
            if (!negotiationInProgress) {
                if (g.phase === g.PHASE.FREE_AGENCY) {
                    ui.updateStatus(g.daysLeft + " days left");
                } else {
                    ui.updateStatus("Idle");
                }
                ui.updatePlayMenu(tx);
            }
        });
    }).then(function () {
        require('../core/league').updateLastDbChange();
    });
}

/**
 * Cancel all ongoing contract negotiations.
 *
 * Currently, the only time there should be multiple ongoing negotiations in the first place is when a user is re-signing players at the end of the season, although that should probably change eventually.
 *
 * @memberOf core.contractNegotiation
 * @param {IDBTransaction} tx An IndexedDB transaction on gameAttributes, messages, and negotiations, readwrite.
 * @return {Promise}
 */
function cancelAll(tx) {
    return tx.negotiations.clear().then(function () {
        require('../core/league').updateLastDbChange();
        ui.updateStatus("Idle");
        return ui.updatePlayMenu(tx);
    });
}

/**
 * Accept the player's offer.
 *
 * If successful, then the team's current roster will be displayed.
 *
 * @memberOf core.contractNegotiation
 * @param {number} pid An integer that must correspond with the player ID of a player in an ongoing negotiation.
 * @return {Promise.<string=>} If an error occurs, resolves to a string error message.
 */
function accept(pid, amount, exp) {
    return Promise.all([
        g.dbl.negotiations.get(pid),
        team.getPayroll(null, g.userTid).get(0)
    ]).spread(function (negotiation, payroll) {
        // If this contract brings team over the salary cap, it's not a minimum;
        // contract, and it's not re-signing a current player, ERROR!
        if (!negotiation.resigning && (payroll + amount > g.salaryCap && amount > g.minContract)) {
            return "This contract would put you over the salary cap. You cannot go over the salary cap to sign free agents to contracts higher than the minimum salary. Either negotiate for a lower contract or cancel the negotiation.";
        }

        // This error is for sanity checking in multi team mode. Need to check for existence of negotiation.tid because it wasn't there originally and I didn't write upgrade code. Can safely get rid of it later.
        if (negotiation.tid !== undefined && negotiation.tid !== g.userTid) {
            return "This negotiation was started by the " + g.teamRegionsCache[negotiation.tid] + " " + g.teamNamesCache[negotiation.tid] + " but you are the " + g.teamRegionsCache[g.userTid] + " " + g.teamNamesCache[g.userTid] + ". Either switch teams or cancel this negotiation.";
        }

        // Adjust to account for in-season signings;
        if (g.phase <= g.PHASE.AFTER_TRADE_DEADLINE) {
            negotiation.player.years -= 1;
        }

        return g.dbl.tx(["players", "playerStats"], "readwrite", function (tx) {
            return tx.players.iterate(pid, function (p) {
                p.tid = g.userTid;
                p.gamesUntilTradable = 15;

                // Handle stats if the season is in progress
                if (g.phase <= g.PHASE.PLAYOFFS) { // Otherwise, not needed until next season
                    p = player.addStatsRow(tx, p, g.phase === g.PHASE.PLAYOFFS);
                }

                p = player.setContract(p, {
                    amount: amount,
                    exp: exp
                }, true);

                if (negotiation.resigning) {
                    eventLog.add(null, {
                        type: "reSigned",
                        text: 'The <a href="' + helpers.leagueUrl(["roster", g.teamAbbrevsCache[g.userTid], g.season]) + '">' + g.teamNamesCache[g.userTid] + '</a> re-signed <a href="' + helpers.leagueUrl(["player", p.pid]) + '">' + p.name + '</a> for ' + helpers.formatCurrency(p.contract.amount / 1000, "M") + '/year through ' + p.contract.exp + '.',
                        showNotification: false,
                        pids: [p.pid],
                        tids: [g.userTid]
                    });
                } else {
                    eventLog.add(null, {
                        type: "freeAgent",
                        text: 'The <a href="' + helpers.leagueUrl(["roster", g.teamAbbrevsCache[g.userTid], g.season]) + '">' + g.teamNamesCache[g.userTid] + '</a> signed <a href="' + helpers.leagueUrl(["player", p.pid]) + '">' + p.name + '</a> for ' + helpers.formatCurrency(p.contract.amount / 1000, "M") + '/year through ' + p.contract.exp + '.',
                        showNotification: false,
                        pids: [p.pid],
                        tids: [g.userTid]
                    });
                }

                return p;
            });
        }).then(function () {
            return cancel(pid);
        }).then(function () {
            require('../core/league').updateLastDbChange();
        });
    });
}

module.exports = {
    accept: accept,
    cancel: cancel,
    cancelAll: cancelAll,
    create: create
};
